src/backend/codegen/include/codegen/base_codegen.h

```diff
@@ -0,0 +1,158 @@
+//
+//  @filename:
+//    base_codegen.h
+//
+//  @doc:
+//    Base class for all the code generators with common implementation
+//
+//---------------------------------------------------------------------------
+#ifndef GPCODEGEN_BASE_CODEGEN_H_  // NOLINT(build/header_guard)
+#define GPCODEGEN_BASE_CODEGEN_H_
+
+#include <string>
+#include <vector>
+
+#include "codegen/utils/codegen_utils.h"
+#include "codegen/codegen_interface.h"
+
+namespace gpcodegen {
+
+/** \addtogroup gpcodegen
+ *  @{
+ */
+
+/**
+ * @brief Base code generator with common implementation that other
+ * 		  code generators can use.
+ *
+ * @tparam FuncPtrType Function type for regular / generated functions.
+ **/
+template<class FuncPtrType>
+class BaseCodeGen: public CodeGenInterface {
+ public:
+  /**
+   * @brief Destroys the code generator and reverts back to using regular
+   * 		version of the generated function.
+   **/
+  virtual ~BaseCodeGen() {
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+  }
+
+  bool GenerateCode(gpcodegen::CodeGenUtils* codegen_utils) final {
+    is_generated_ = DoCodeGeneration(codegen_utils);
+    return is_generated_;
+  }
+
+  bool SetToRegular() final {
+    assert(nullptr != regular_func_ptr_);
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+    return true;
+  }
+
+  bool SetToGenerated(gpcodegen::CodeGenUtils* codegen_utils) final {
+    if (false == IsGenerated()) {
+      assert(*ptr_to_chosen_func_ptr_ == regular_func_ptr_);
+      return false;
+    }
+
+    auto compiled_func_ptr = codegen_utils->GetFunctionPointerTypeDef<
+        FuncPtrType>(GetUniqueFuncName());
+
+    if (nullptr != compiled_func_ptr) {
+      *ptr_to_chosen_func_ptr_ = compiled_func_ptr;
+      return true;
+    }
+    return false;
+  }
+
+  void Reset() final {
+    SetToRegular();
+  }
+
+  const std::string& GetOrigFuncName() const final {
+    return orig_func_name_;
+  }
+
+  const std::string& GetUniqueFuncName() const final {
+    return unique_func_name_;
+  }
+
+  bool IsGenerated() const final {
+    return is_generated_;
+  }
+
+  /**
+   * @return Regular version of the corresponding generated function.
+   *
+   **/
+  FuncPtrType GetRegularFuncPointer() {
+    return regular_func_ptr_;
+  }
+
+  /**
+   * @brief Sets up the caller to use the corresponding regular version of the
+   *        generated function.
+   *
+   * @param regular_func_ptr       Regular version of the generated function.
+   * @param ptr_to_chosen_func_ptr Reference to caller.
+   *
+   * @return true on setting to regular version.
+   **/
+  static bool SetToRegular(FuncPtrType regular_func_ptr,
+                           FuncPtrType* ptr_to_chosen_func_ptr) {
+    *ptr_to_chosen_func_ptr = regular_func_ptr;
+    return true;
+  }
+
+ protected:
+  /**
+   * @brief Constructor
+   *
+   * @param orig_func_name         Original function name.
+   * @param regular_func_ptr       Regular version of the generated function.
```

Comment:
@gcaragea  Are you talking about this comment "Regular version of the generated function". This will change to something like "Regular version of the target function" ? I like this to call as target function :)